src/memcard_simulator_sm.c

```diff
@@ -0,0 +1,325 @@
+#include <complex.h>
+#include <sys/cdefs.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include "memcard_simulator_sm.h"
+#include "pico/stdlib.h"
+#include "hardware/pio.h"
+#include "hardware/irq.h"
+#include "bsp/board.h"
+#include "psxSPI.pio.h"
+#include "memory_card.h"
+
+#define PIN_DAT 5
+#define PIN_CMD 6
+#define PIN_SEL 7
+#define PIN_CLK 8
+#define PIN_ACK 9
+
+#define MEMCARD_WAKE 0x81
+#define MEMCARD_READ 0x52
+#define MEMCARD_WRITE 0x57
+#define MEMCARD_ID 0x53
+
+#define IDLE_TIMEOUT_BEFORE_SYNC 1000
+
+enum memcard_states {
+    MC_IDLE = 1,
+    MC_COMMAND,
+    MC_SEND_ID,
+    MC_RECV_ADDR,
+    MC_EXECUTE_READ,
+    MC_EXECUTE_WRITE,
+    MC_EXECUTE_ID,
+    MC_ABORT,
+    MC_END
+};
+
+uint8_t current_state = MC_IDLE;
+uint8_t next_state = MC_IDLE;
+uint8_t command_state = MC_IDLE;
+uint8_t checksum = 0x00;
+uint8_t recv_checksum = 0x00;
+uint8_t sm_byte_counter = 0;
+uint8_t sm_address = 0x00;
+uint8_t id_data[] = {MC_ACK1, MC_ACK2, 0x04, 0x00, 0x00, 0x80};
+bool in_transaction = false;
+
+PIO pio = pio0;
+
+uint smSelMonitor;
+uint smCmdReader;
+uint smAckSender;
+uint smDatWriter;
+
+uint offsetSelMonitor;
+uint offsetCmdReader;
+uint offsetDatWriter;
+uint offsetAckSender;
+
+MemoryCard mc;
+
+void pio0_irq0() {
+    pio_set_sm_mask_enabled(pio, 1 << smCmdReader | 1 << smDatWriter, false);
+    pio_restart_sm_mask(pio, 1 << smCmdReader | 1 << smDatWriter);
+    pio_sm_exec(pio, smCmdReader, pio_encode_jmp(offsetCmdReader));	// restart smCmdReader PC
+    pio_sm_exec(pio, smDatWriter, pio_encode_jmp(offsetDatWriter));	// restart smDatWriter PC
+    pio_sm_clear_fifos(pio, smCmdReader);
+    pio_sm_clear_fifos(pio, smDatWriter);
+    pio_interrupt_clear(pio0, 0);
+    pio_enable_sm_mask_in_sync(pio, 1 << smCmdReader | 1 << smDatWriter);
+
+    // Reset mc state machine
+    current_state = MC_IDLE;
+    next_state = MC_IDLE;
+    command_state = MC_IDLE;
+    sm_byte_counter = 0;
+    sm_address = 0x00;
```

Comment:
```suggestion
    sm_address = 0x0000;
```
For consistency to highlight the size

Reply:
Done